common/functions/src/scalars/expressions/cast.rs

```diff
@@ -67,7 +67,12 @@ impl Function for CastFunction {
     }
 
     fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
-        if columns[0].data_type() == &self.cast_type {
+        // TODO: ideally we don't need to do remove nullable -- the input should be always non-nullable.
+        // The null value will be masked after function.eval.
+        let source_type = columns[0].data_type().remove_nullable();
```

Comment:
So TODO will be addressed in another pr?

Reply:
Right, the PR just introduce the new type, doing the best to behavior same as it is now. It doesn't refactor any return_type or eval. 

The feature we talked, like
1.  moving nullable check out of return_type 
2. pass_through_null returns multiple state,
3. remove DataTypeWithNull
Will be addressed in another PR.